### PR TITLE
Smear

### PR DESCRIPTION
**Pull request by irischang020**

### CODE REVIEW

File: pysingfel/detector/autoranging.py [hunk 1]
```diff
@@ -26,36 +26,38 @@ def setupMatrices(self):
         self.offsets = self.matricize(self.offsets)
         self.switchPoints = self.matricize(self.switchPoints, None, self.switchPointVariations)
 
-    def matricize(self, array, relativeSmear=None, absoluteSmear=None): # better to separate out smearing function
+    def matricize(self, array, relativeSmear=None, absoluteSmear=None):
```
Comment (chuckie82) [inline pysingfel/detector/autoranging.py:29]: a function should do one thing. here it does two things, that's why we want to separate this out.
def matricize(self, array)
and
def smear(array or matrix, relative, absolute)

Comment (irischang020) [reply pysingfel/detector/autoranging.py]: def matricize(self, array)
and 
def smear(self, matrices, relativeSmear=None, absoluteSmear=None)

**Comment by irischang020** [comment]
https://github.com/chuckie82/pysingfel/pull/83/files

**Comment by AntoineDujardin** [comment]
The smear function looks like 2 functions: smear_absolute and smear_relative.
Should we consider splitting them? Is there a use-case for doing the 2 kinds of smear on the same array?

**Comment by irischang020** [comment]
1. Not enough examples in Philip's code to know if there's a use-case for doing the 2 kinds of smear on the same matices. 
2. relativeSmear and absoluteSmear are both under smear method so I don't think we need to split them.

File: pysingfel/detector/autoranging.py [hunk 2]
```diff
@@ -10,54 +10,55 @@ def __init__(self, cameraConfig, *args, **kwargs):
...
     def updateFlatSwitchPoints(self, flatSwitchPoints):
-        self.switchPoints = self.matricize(flatSwitchPoints)
+        self.switchPoints = self.smear(self.matricize(flatSwitchPoints), None, None)
 
     def updateSwitchPoints(self, switchPoints):
         self.switchPoints = switchPoints
 
     def setupMatrices(self):
-        self.residualGains = self.matricize(self.gains, self.gainErrors, None)/self.matricize(self.gains, None, None) # Q: ADU/keV?
+        self.residualGains = self.smear(self.matricize(self.gains), self.gainErrors, None)/self.smear(self.matricize(self.gains), None, None) # Q: ADU/keV?
```
Comment (chuckie82) [inline pysingfel/detector/autoranging.py:25]: Inefficient to call matricize(gains) twice here.
Actually, the denominator is wrong. It should just be matricize()?
Please check whether the output is still the same as the original code.

Comment (irischang020) [reply pysingfel/detector/autoranging.py:25]: I fixed the particle orientation and printed out the calibrated photons at pixel[10][0][383].
('calib_photons[10][0][383]: ', 521.7391304347826) --> The output is still the same as the original code.
Screen Shot 2020-07-17 at 7 34 22 PM

Comment (irischang020) [reply pysingfel/detector/autoranging.py:25]: The return value of `matricize()` is tmp (type:list, len=nRanges). Pass it into `smear` and apply relativeSmear/absoluteSmear then the return value is np.array(tmp) (type:array, size=detector dimension).

Comment (chuckie82) [reply pysingfel/detector/autoranging.py]: matrice() wasn't returning an array but a list

Comment (irischang020) [reply pysingfel/detector/autoranging.py]: I ran the example script `ExampleAutoranging.py` with the updated `autoranging.py` but got the following error:
(ana-1.5.12) [iris@psanagpu109 scripts]$ python ExampleAutoranging.py 
/reg/neh/home/iris/Software/pysingfel/pysingfel/detector/autoranging.py:25: RuntimeWarning: invalid value encountered in divide
  self.residualGains = self.smear(self.matricize(self.gains), self.gainErrors, None)/self.matricize(self.gains) # unitless
/reg/neh/home/iris/Software/pysingfel/pysingfel/build_autoranging_frames.py:43: RuntimeWarning: invalid value encountered in greater
  self.frame[np.where(self.frame > self.det.maxEnergy)] = self.det.maxEnergy
('calib_photons[10][0][383]: ', nan)
Traceback (most recent call last):
  File "ExampleAutoranging.py", line 48, in <module>
    viz.imshow(img)
  File "/reg/neh/home/iris/Software/pysingfel/pysingfel/visualizer.py", line 47, in imshow
    plt.colorbar()
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/pyplot.py", line 2320, in colorbar
    ret = gcf().colorbar(mappable, cax = cax, ax=ax, **kw)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/figure.py", line 2098, in colorbar
    cb = cbar.colorbar_factory(cax, mappable, **cb_kw)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 1399, in colorbar_factory
    cb = Colorbar(cax, mappable, **kwargs)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 945, in __init__
    ColorbarBase.__init__(self, ax, **kw)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 327, in __init__
    self.draw_all()
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 353, in draw_all
    self._config_axes(X, Y)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 461, in _config_axes
    self.update_ticks()
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 378, in update_ticks
    ticks, ticklabels, offset_string = self._ticker()
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 622, in _ticker
    b = np.array(locator())
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/ticker.py", line 2098, in __call__
    return self.tick_values(vmin, vmax)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/ticker.py", line 2124, in tick_values
    "Data has no positive values, and therefore can not be "
ValueError: Data has no positive values, and therefore can not be log-scaled.

**Comment by chuckie82** [comment]
@irischang020 I think there was an indentation error with the absolute smear function in matrices += smears. 
matrices would be smeared twice when absoluteSmear is not a scalar.
Please check whether my correction is right or not.

This is a good example of when we need a unit test to test both when absolute smear is 1) a scalar and 2) not a scalar

**Comment by irischang020** [comment]
@chuckie82 Yes I also found that there was an indentation error with the absolute smear method in `smear(self, matrices, relativeSmear=None, absoluteSmear=None)`.

The example script `ExampleAutoranging.py` with the updated `autoranging.py` splitting relativeSmear and absoluteSmear gave the same output as the original code.

**Comment by chuckie82** [comment]
@irischang020 thanks for checking with original Phil's code. 
please put in unit tests for a flat field matrix with known smear value and just check one pixel value is correct. you can hard-code the pixel-value answer in the unit tests. (let me know if I'm not explaining this clearly)
So 4 tests for every combination of relative, absolute, scalar, non-scalar cases

**Comment by irischang020** [comment]
There's np.random.normal in smear. How do we set a seed

**Comment by irischang020** [comment]
`np.random.normal()` gets called in `smear`. To check if a certain pixel value is consistent with that in the unit tests, we'll need to set a seed in the function. How do we do that?

**Comment by irischang020** [comment]
I passed in a flat field matrix and found that the certain pixel value was inconsistent with that in the unit tests. Then I checked the three versions of `autoranging.py` and `build_autoranging_frames.py` but found that all were inconsistent.
Then I realized that np.random.normal() gets called in smear. To check if a certain pixel value is consistent with that in the unit tests, we'll need to set a seed in the function. How do we do that?

**Comment by AntoineDujardin** [comment]
Call `numpy.random.seed()` in the test, before calling the function you want to test. That'll make sure you're always getting the same answer.

**Comment by irischang020** [comment]
I called `numpy.random.seed(0)` before `BuildAutoRangeFrames` but the results were different.
~iris/Software/pysingfel/examples/scripts/ExampleAutoranging_flatfield.py

**Comment by irischang020** [comment]
The three versions of `autoranging.py` and `build_autoranging_frames.py` give the same result as the original code for a certain pixel value if setting numpy.random.seed() before `np.random.normal()` in `smear`. --> a sanity check that `autoranging.py` and `build_autoranging_frames.py` are correct.

**Comment by chuckie82** [comment]
ok, let's convert your sanity checks to unit tests and I'll merge this PR

**Comment by chuckie82** [comment]
I recommend using assert np.isclose() next time rather than ==
I'll merge now. thanks.

**Pull request merged**